app/src/main/java/com/nutomic/syncthingandroid/util/ConfigXml.java

```diff
@@ -110,9 +124,38 @@ public String getUserName() {
      */
     @SuppressWarnings("SdCardPath")
     public void updateIfNeeded() {
-        Log.i(TAG, "Checking for needed config updates");
         boolean changed = false;
+
+        /* Perform one-time migration tasks on syncthing's config file when coming from an older config version. */
+        changed = migrateSyncthingOptions() || changed;
+
+        /* Get refs to important config objects */
         NodeList folders = mConfig.getDocumentElement().getElementsByTagName("folder");
+
+        /**
+          * Force-disable fsWatcher for all folders if prefUseFolderObserver has been manually set
+          * in experimental options. This should give users the option to go back to the legacy
+          * implementation with FolderObserver watching for file changes instead of fsWatcher.
+          * Intended to be advised in the issue tracker if a user encounters a critical bug with
+          * the new fsWatcher. To be removed in a future release.
+        */
+        /* Get preference - PREF_USE_FOLDER_OBSERVER */
+        mPreferences = PreferenceManager.getDefaultSharedPreferences(mContext);
+        boolean prefUseFolderObserver = mPreferences.getBoolean(Constants.PREF_USE_FOLDER_OBSERVER, false);
+        if (prefUseFolderObserver) {
+            Log.i(TAG, "Disabling fsWatcher on all folders because experimental option to use FolderObserver was manually set.");
+            for (int i = 0; i < folders.getLength(); i++) {
+                Element r = (Element) folders.item(i);
+
+                // Disable "fsWatcherEnabled" attribute.
+                if (r.hasAttribute("fsWatcherEnabled")) {
+                    r.setAttribute("fsWatcherEnabled", Boolean.toString(false));
```

Comment:
Don't we want to enable it if the internal watcher is disabled?

Because it seems that toggling the `prefUseFolderObserver` post config migration will lead to a state where the syncthings own watcher cannot be enabled under any circumstances via the app.

Reply:
If it's that complicated, I would just remove FolderObserver.java, always set fsWatcherEnabled and release it as a beta. If there are any problems, people will report them. The Go version is definitely better tested, so I dont think there will be any problems.

Reply:
Yeah perhaps just remove the built in watcher.

Reply:
Ok, I will do some more testing on my phone with fsWatcherEnabled as its running fine since I installed the rc4 debug release. If there arent any problems will remove it. Would you suggest disabling the code (dont load it and make it hidden) or should I completely remove FolderObserver.java and relevant code out of the project?

Reply:
Just remove the code, we have version control for history.

Reply:
Just remove the code, we have version control for history.

Reply:
Done with commit https://github.com/syncthing/syncthing-android/pull/1051/commits/8d0a1dc0e36a3a99b95969f12de2d193b0f57377